x-pack/plugins/security_solution/common/api/detection_engine/prebuilt_rules/perform_rule_upgrade/perform_rule_upgrade_route.schema.yaml

```diff
@@ -0,0 +1,280 @@
+              $ref: '#/components/schemas/FieldUpgradeRequest'
+            building_block_type:
+              $ref: '#/components/schemas/FieldUpgradeRequest'
+            from:
+              $ref: '#/components/schemas/FieldUpgradeRequest'
+            interval:
+              $ref: '#/components/schemas/FieldUpgradeRequest'
+            exceptions_list:
+              $ref: '#/components/schemas/FieldUpgradeRequest'
+            rule_name_override:
+              $ref: '#/components/schemas/FieldUpgradeRequest'
+            timestamp_override:
+              $ref: '#/components/schemas/FieldUpgradeRequest'
+            timestamp_override_fallback_disabled:
+              $ref: '#/components/schemas/FieldUpgradeRequest'
+            timeline_id:
+              $ref: '#/components/schemas/FieldUpgradeRequest'
+            timeline_title:
+              $ref: '#/components/schemas/FieldUpgradeRequest'
+            index:
+              $ref: '#/components/schemas/FieldUpgradeRequest'
+            data_view_id:
+              $ref: '#/components/schemas/FieldUpgradeRequest'
+            query:
+              $ref: '#/components/schemas/FieldUpgradeRequest'
+            language:
+              $ref: '#/components/schemas/FieldUpgradeRequest'
+            filters:
+              $ref: '#/components/schemas/FieldUpgradeRequest'
+            saved_id:
+              $ref: '#/components/schemas/FieldUpgradeRequest'
+            machine_learning_job_id:
+              $ref: '#/components/schemas/FieldUpgradeRequest'
+            anomaly_threshold:
+              $ref: '#/components/schemas/FieldUpgradeRequest'
+            threat_query:
+              $ref: '#/components/schemas/FieldUpgradeRequest'
+            threat_mapping:
+              $ref: '#/components/schemas/FieldUpgradeRequest'
+            threat_index:
+              $ref: '#/components/schemas/FieldUpgradeRequest'
+            threat_filters:
+              $ref: '#/components/schemas/FieldUpgradeRequest'
+            threat_indicator_path:
+              $ref: '#/components/schemas/FieldUpgradeRequest'
+            threat_language:
+              $ref: '#/components/schemas/FieldUpgradeRequest'
+            new_terms_fields:
+              $ref: '#/components/schemas/FieldUpgradeRequest'
+            history_window_start:
+              $ref: '#/components/schemas/FieldUpgradeRequest'
+
+    FieldUpgradeRequest:
+      type: object
+      required:
+        - pick_version
+      properties:
+        pick_version:
+          type: string
+          enum: [BASE, CURRENT, TARGET, MERGED, RESOLVED]
+        resolved_value:
+          oneOf:
+            - $ref: '../../model/rule_schema/common_attributes.schema.yaml#/components/schemas/RuleName'
+            - $ref: '../../model/rule_schema/common_attributes.schema.yaml#/components/schemas/RuleTagArray'
+            - $ref: '../../model/rule_schema/common_attributes.schema.yaml#/components/schemas/RuleDescription'
```

Comment:
What is the reason for making this type a union? The `ruleName` prop should have the `resolved_value` matching the `ruleName` type, not any of these. Can it be implemented that way?

Reply:
I'm trying to reverse engineer these types. Notice the generic `FieldUpgradeRequest` type, which is what I tried to replicate with a union:
```ts
export interface PerformRuleUpgradeRequestBody {
  mode: 'ALL_RULES' | 'SPECIFIC_RULES';
  pick_version?: 'BASE' | 'CURRENT' | 'TARGET' | 'MERGED';
  rules: SingleRuleUpgradeRequest[]; // required if mode is SPECIFIC_RULES
}

export interface SingleRuleUpgradeRequest {
  id: RuleObjectId;
  pick_version?: 'BASE' | 'CURRENT' | 'TARGET' | 'MERGED';
  fields?: {
    name?: FieldUpgradeRequest<RuleName>;
    description?: FieldUpgradeRequest<RuleDescription>;
    // etc
    // Every non-specified field will default to pick_version: 'MERGED'.
    // If pick_version is MERGED and there's a merge conflict the endpoint will throw.
  };
  rule_revision: number;
}

export interface FieldUpgradeRequest<T> {
  pick_version: 'BASE' | 'CURRENT' | 'TARGET' | 'MERGED' | 'RESOLVED';
  resolved_value: T; // required if pick_version is RESOLVED; type depends on the rule field type
}
```

So each of the properties within `fields` in `SingleRuleUpgradeRequest` should be of type `FieldUpgradeRequest` because each property should be able to be updated to any of the `pick_version`s listed in the full enum (which includes `RESOLVED`).

**Only if** the `pick_version` value for that field is `RESOLVED`, then the `resolved_value` can take the value of the prop (`RuleName` for the case of the `name` property).

I'm not sure how to write the OAS in some other way without a union.

Reply:
Maybe instead of having a "generic" `FieldUpgradeRequest` we could have something like `RuleNameFieldUpgradeRequest`, which directly `ref`s to `RuleName`. And create a separate schema for each of the updatable fields. 

With such an approach we would need a union, but it would mean a lot more schemas manually created.

What do you think>

Reply:
I think we need to list every upgradeable field, something like this:

```yaml
RuleUpgradeSpecifier:
  type: object
  properties:
    fields:
      type: object
      properties:
        name:
          type: object
          required:
            - pick_version
          properties:
            pick_version:
              $ref: '#/components/schemas/FieldPickVersionValues'
            resolved_value:
              $ref: '../../model/rule_schema/common_attributes.schema.yaml#/components/schemas/RuleName'
        tags:
          type: object
          required:
            - pick_version
          properties:
            pick_version:
              $ref: '#/components/schemas/FieldPickVersionValues'
            resolved_value:
              $ref: '../../model/rule_schema/common_attributes.schema.yaml#/components/schemas/RuleTagArray'
```

This approach is a bit more verbose than using a union with all possible values, but it will produce the most accurate type. I believe the effort is worth it.